libs/vkd3d/swapchain.c

```diff
@@ -1742,13 +1721,24 @@ static VkResult d3d12_swapchain_queue_present(struct d3d12_swapchain *swapchain,
     }
 
     present_info.waitSemaphoreCount = 1;
-    present_info.pWaitSemaphores = &swapchain->vk_semaphores[swapchain->vk_image_index];
+    present_info.pWaitSemaphores = &swapchain->vk_present_semaphores[swapchain->vk_image_index];
 
     if ((vr = vk_procs->vkQueuePresentKHR(vk_queue, &present_info)) >= 0)
     {
-        swapchain->vk_image_index = INVALID_VK_IMAGE_INDEX;
+        swapchain->frame_id = (swapchain->frame_id + 1) % swapchain->buffer_count;
+
         /* Could get SUBOPTIMAL here. Deal with it later. */
         vr = VK_SUCCESS;
+
+        /* Try to acquire our next image to avoid blocking next frame. */
+        if ((vr = vk_procs->vkAcquireNextImageKHR(vk_device, swapchain->vk_swapchain, UINT64_MAX,
+                    swapchain->vk_acquire_semaphores[swapchain->frame_id], VK_NULL_HANDLE, &swapchain->vk_image_index)))
+        {
+            /* Didn't manage to get a non-suboptimal image or any at all.
+             * The last present we did was a success so don't remake the
+             * swapchain now. Retry again at the next presentation. */
```

Comment:
Aren't you leaving the semaphore signaled here in case of `VK_SUBOPTIMAL_KHR`, without waiting for it? And if you succeed next time, eventually you will reuse your signaled semaphore, which is invalid.

Reply:
Yes. I guess I should just recreate in this case as well to avoid that.

Reply:
Same comment as above.